### PR TITLE
#75 zipfiles

### LINKED ISSUE #75: Add support for zipfiles
Related with " Add support to handle more than one file #30 "

### PR DESCRIPTION
**Pull request by SbastianGarzon**
Closes #75 

This pull request includes:
1. Support for zipfiles and folders (directories) - (recursive)
- Bbox merge (output : EPSG 4326 )
- Tbox merge (output : yyyy-mm-dd)
2. Functions to merge bounding box and time box.
3. Standardized output formats for different type of files and results (To facilitate merging of bbox and tbox from multiple files)
4. Test for folders (directories) and zipfiles
5. Documentation update

### CODE REVIEW

File: tests/test_api.py [hunk 1]
```diff
@@ -15,76 +16,127 @@ def test_defaults():
...
+    assert "tbox" in result
+    assert result["bbox"] == [7.60168075561523, 34.7, 142.0, 51.9746240298775]
+    assert result["tbox"] == ['2017-04-08', '2020-02-06']
+
+def test_empty_zipfile():
+    folder_name = "tests/testdata/folders/temp_empty_folder"
+    os.mkdir(folder_name)
+    hf.create_zip(folder_name)
+    result = geoextent.fromDirectory(folder_name+'.zip', bbox=True, tbox=True)
+    os.remove(folder_name + '.zip')
```
Comment (nuest) [inline tests/test_api.py:79]: If there is an error in the line before, the `os.remove(..)` will not be called and the test will leave a file artefact.

Please take advantage of the `with ...` statement to ensure the files created for the tests are cleaned up, see https://docs.python.org/3.4/library/tempfile.html#examples

Maybe something like

```python
with tempfile.TemporaryFile(suffix = 'zip') as zipfile:
    create_zip(zipfile, folder_name)
    result = geoextent.fromDirectory(zipfile, bbox=True, tbox=True)
    assert "bbox" not in result
    assert "tbox" not in result
```

with a `create_zip()` function that also takes a target file name - and that's it.

File: tests/testdata/data-sources.md [hunk 2]
```diff
@@ -16,6 +16,7 @@
 | [geoextent]         | onePoint                          | geojson   | PDDL                               |
 | [geoextent]         | invalid_coordinate                | geojson   | PDDL                               |
 | [geoextent]         | onePoint                          | geojson   | PDDL                               |
+| [Carto BCN / martgnz]         | Barcelona districts                          | geojson   | CC-BY 4.0                              |
```
Comment (nuest) [inline tests/testdata/data-sources.md:19]: Nice touch - Barcelona! :woman_dancing:

File: tests/test_cli.py [hunk 3]
```diff
@@ -237,14 +237,14 @@ def test_multiple_files(script_runner):
     assert "[6.574722, 51.434444, 4.3175, 53.217222]" in ret.stdout, "bboxes and time values of all files inside folder, are printed to console"
     assert "[292063.81225905, 5618144.09259115, 302531.3161606, 5631223.82854667]" in ret.stdout, "bboxes and time values of all files inside folder, are printed to console"
 
-@pytest.mark.skip(reason="director input not implemented yet")
 def test_folder(script_runner):
```
Comment (nuest) [inline tests/test_cli.py:240]: What about a CLI test for a ZIP file?

**Review by nuest — CHANGES_REQUESTED** [review]
Good job - a few wording comments and requests for a bit more verbose logging. And I think a test case is missing and temp files in tests can be handled more elegantly.

File: geoextent/lib/helpfunctions.py [hunk 4]
```diff
@@ -193,3 +193,114 @@ def date_parser(datetime_list, num_sample=None):
...
+                box.AddPoint(bbox[0][2], bbox[0][3])
+                box.AddPoint(bbox[0][0], bbox[0][3])
+                box.AddPoint(bbox[0][0], bbox[0][1])
+
+                if bbox[1] != "4326":
+                    logger.warning("Transforming bbox CRS from EPSG [{}] to [{}] ".format(bbox[1], WGS84_EPSG_ID))
+                    source = osr.SpatialReference()
+                    source.ImportFromEPSG(int(bbox[1]))
+                    transform = osr.CoordinateTransformation(source, des_srs)
+                    box.Transform(transform)
```
Comment (nuest) [inline geoextent/lib/helpfunctions.py:262]: Will this transformation never fail or throw an error?

File: docs/source/howto/api.rst [hunk 5]
```diff
@@ -81,3 +81,45 @@ Output:
 (`source of file muenster_ring_zeit.geojson`_)
 
 .. _source of file muenster_ring_zeit.geojson: https://github.com/o2r-project/geoextent/blob/master/tests/testdata/geojson/muenster_ring_zeit.geojson
+
+
+Folders or zipfiles
```
[the remark this answers is not in the document]
Comment (SbastianGarzon) [reply docs/source/howto/api.rst:86]: Done !

File: docs/source/howto/api.rst [hunk 6]
```diff
@@ -81,3 +81,45 @@ Output:
...
+-------------------
+
+Geoextent also supports queries for multiple files inside ``folders`` or ``zipfiles``.
```
[the remark this answers is not in the document]
Comment (SbastianGarzon) [reply docs/source/howto/api.rst:89]: Done !

File: docs/source/howto/api.rst [hunk 7]
```diff
@@ -81,3 +81,45 @@ Output:
...
+Folders or zipfiles
+-------------------
+
+Geoextent also supports queries for multiple files inside ``folders`` or ``zipfiles``. 
+
+::
+
+   geoextent.fromDirectory(input, bbox, time)
+
+It takes raw data (in the form of a string, boolean, boolean) to output result based on these fields.
```
[the remark this answers is not in the document]
Comment (SbastianGarzon) [reply docs/source/howto/api.rst:95]: This is the standard message in the documentation (see Individual files), I think it is not required as the content is better described in the list below.

File: geoextent/lib/extent.py [hunk 8]
```diff
@@ -40,6 +42,63 @@ def fromDirectory(path, bbox=False, tbox=False):
...
+        else:
+            logger.info("Inspecting folder {}, is zip ? {}".format(filename, str(isZip)))
+            if os.path.isdir(os.path.join(path,filename)):
+                metadata_directory[filename] = fromDirectory(os.path.join(path, filename), bbox, tbox)
+            else:
+                metadata_file = fromFile(os.path.join(path, filename), bbox, tbox)
+                metadata_directory[str(filename)] = metadata_file
+
+    if isZip:
+        file_format = 'zipfile'
```
[the remark this answers is not in the document]
Comment (SbastianGarzon) [reply geoextent/lib/extent.py:78]: Other format names do not include "file" so I'll change it to 'zip'.

File: geoextent/lib/helpfunctions.py [hunk 9]
```diff
@@ -193,3 +193,114 @@ def date_parser(datetime_list, num_sample=None):
...
+                boxes.append(y['tbox'][0])
+                boxes.append(y['tbox'][1])
+            except:
+                pass
+
+    if len(boxes) == 0:
+        time_ext = None
+
+    elif len(boxes) >= 1:
+        logger.info("Boxes to compare {}".format(boxes))
```
[the remark this answers is not in the document]
Comment (SbastianGarzon) [reply geoextent/lib/helpfunctions.py:296]: Ups ! I'll also change the message to "Time boxes to compare from: {}"

File: geoextent/lib/helpfunctions.py (hunk 4 above)
Comment (SbastianGarzon) [reply geoextent/lib/helpfunctions.py:262]: I'm using a "long" try except for this part. The most likely error in here is that the detected CRS name does not exist (Line 260). If there is an error creating the points or during the transformation, that geographical extent (CRS+BBOX) is ignored for the merging computation (no points are added to the 'merged' geometry). I'll add a debug message in the except statement to make it clear the specific error.

File: tests/test_cli.py (hunk 3 above)
Comment (SbastianGarzon) [reply tests/test_cli.py:240]: I'll include  a CLI test for ZIP files !

File: geoextent/lib/extent.py [hunk 10]
```diff
@@ -40,6 +42,63 @@ def fromDirectory(path, bbox=False, tbox=False):
...
+            metadata_directory[filename] = fromDirectory(os.path.join(path,filename),bbox,tbox)
+        else:
+            logger.info("Inspecting folder {}, is zip ? {}".format(filename, str(isZip)))
+            if os.path.isdir(os.path.join(path,filename)):
+                metadata_directory[filename] = fromDirectory(os.path.join(path, filename), bbox, tbox)
+            else:
+                metadata_file = fromFile(os.path.join(path, filename), bbox, tbox)
+                metadata_directory[str(filename)] = metadata_file
+
+    if isZip:
```
[the remark this answers is not in the document]
Comment (SbastianGarzon) [reply geoextent/lib/extent.py:77]: Done !

File: geoextent/lib/helpfunctions.py [hunk 11]
```diff
@@ -193,3 +193,114 @@ def date_parser(datetime_list, num_sample=None):
...
+
+    abs_path = os.path.abspath(zippedFile)
+    root_folder = os.path.split(abs_path)[0]
+    zip_name = os.path.split(abs_path)[1][:-4]
+    zip_folder_path = os.path.join(root_folder, zip_name)
+
+    with zipfile.ZipFile(abs_path) as zipf:
+        zipf.extractall(zip_folder_path)
+
+def create_zip(folder):
```
[the remark this answers is not in the document]
Comment (SbastianGarzon) [reply geoextent/lib/helpfunctions.py:211]: Yeah ! That help function is only needed for the tests. I'll move it to the test directory !

File: geoextent/lib/helpfunctions.py (hunk 4 above)
Comment (nuest) [reply geoextent/lib/helpfunctions.py:262]: Just to clarify: `.ImportFromEPSG` throws errors, but `.CoordinateTransformation` and `.Transform` do never throw errors but just return an empty geometry?

**Review by nuest — APPROVED** [review]

**Pull request merged**